Full support for High DPI displays on Windows, Mac and Linux

AntonyBoucher opened the pull request:
Tiles (fonts and graphics) are regenerated from a high-resolution image every time the window is resized. This allows us to perfectly fit the scene into the window, leaving no black padding on the sides.

To access all pixels in a HiDPI setting, we also switch from rendering surfaces with `SDL_UpdateWindowSurface`, to rendering textures with `SDL_RenderPresent`.

The window is a mosaic of tiles which are either `N` or `N+1` pixels wide, and either `M` or `M+1` pixels tall. The large PNG is loaded once into a surface. When the window size changes, 4 textures are regenerated with copies of it, downscaled to the four possible combinations of tile width and height. Each tile gets scaled down separately.

The downscaling is performed in linear color space, rather than in gamma-compressed space which causes dimming (average of a black pixel and a white pixel should be 50% grey = 0.5 luminance = 187 sRGB, not 128). See [this blog post](https://blog.johnnovak.net/2016/09/21/what-every-coder-should-know-about-gamma/#image-resizing) for rationale. For performance, we assume a gamma of 2.0.

The tile is logically split into 16 parts (sliced by 3 vertical lines and 3 horizontal ones), so that:

- the four corners can be positioned to preserve the tile's aspect ratio, possibly filling the blank spaces at the top and bottom;
- the top, bottom, left, right, and center areas can be independently aligned with output pixels to improve sharpness.

Some tiles, like walls and doors, always stretch to fill the space regardless of aspect ratio. Other tiles must preserve aspect ratio, but we allow up to 20% stretch for graphic tiles and 40% for text tiles.

For text tiles (letters, digits, punctuation signs), the characters' x-height and baseline (corresponding to the top and bottom of "x") are pixel-aligned regardless of outlines. This ensures that all letters neatly line up without jumping up and down. We also reduce perceived boldness by applying a custom brightness curve; it makes the text more legible at the smaller sizes.

Optimum pixel alignment is discovered by brute-force, trying all possible values in 0.1 pixel increment (up to 0.5 pixel). Because this is a time-consuming process, it is done only once and saved to disk (`tiles.bin`, committed to git along `tiles.png`). If you ever modify `tiles.png`, you should regenerate `tiles.bin`: simply delete the existing file then restart Brogue.

Wall tops are diagonal sine waves, approximately 4 pixels apart.

If the floor tile is completely blank, we fill it with dust-like noise (isolated dots). This isn't needed for the official Brogue CE tiles, but gives a great, resolution-dependent solution when using custom tiles without the Hop-o'-My-Thumb / Pacman-inspired floor markers.

If the window is so large that tiles need to be over 64x64 pixels, we generate a single, large texture instead of four and use it for all tiles, allowing the renderer to do some linear interpolation.

To ensure compatibility with older OpenGL drivers, texture dimensions are always powers of 2.

OpenGL drivers don't like alternating between different textures too much, so we first draw the background colors then do 4 passes over the tiles, one per texture.

Some video drivers are quite inefficient, notably in virtual machines, so there is a new `--no-gpu` command-line parameter to disable hardware acceleration. The software renderer does not support HiDPI, though.

AntonyBoucher commented:
I fixed the two problems reported to me so far:
- High DPI not effective on Windows with 2 monitors, one standard and one hi-dpi
- Switching to another app and back leaves the full-screen mode

tmewett reviewed (commented):
Looks great and works really well. Great job! Some comments below.

Maybe I'm being a bit dense, but I haven't reviewed some parts in detail because it's a bit hard for me to work out what's going on. The code clarity is generally excellent, but I think we need some overview documentation on

* getPadding (i.e. vertical, bounded at 1/4)
* prepareTile
* optimizeTiles
* tileShifts

There's good info the commit message that would be good to put in and expand on in the code.

AntonyBoucher commented on src/platform/tiles.c:
```diff
@@ -0,0 +1,714 @@
+
+    // accumulator for pixel values (linear color space), encoded as
+    // 0xCCCCCCCCSSSSSSSS where C is a counter and S is a sum of squares
+    uint64_t *values = malloc(tileWidth * tileHeight * 8);
```
OK

AntonyBoucher commented on src/platform/tiles.c:
```diff
@@ -0,0 +1,714 @@
+
+    loadTiles(renderer, outputWidth, outputHeight);
+
+    if (!softwareRendering) {
+        // black out the frame (double-buffering invalidated it)
+        if (SDL_SetRenderDrawColor(renderer, 0, 0, 0, 0) < 0) sdlfatal(__FILE__, __LINE__);
+        if (SDL_RenderClear(renderer) < 0) sdlfatal(__FILE__, __LINE__);
+    }
```
When using SDL's software renderer, either explicitly with  `--no-gpu` or because the machine doesn't offer hardware acceleration, this optimization drops the CPU from 10% down to 2%, so I'd stay it is essential to have it.

This behavior of the software renderer is indeed not specified in the SDL documentation, but it is true, at least in SDL 2.0.12. The software renderer draws onto a single surface (no double-buffering), which is never invalidated nor modified except through the drawing functions. That's why it works, on all platforms. If this undocumented behavior changes in a future version we can adapt, but until then keeping the CPU as low as possible is a must-have on battery-powered devices.

tmewett replied:
Ok great, good point. I agree it makes sense to assume this, as it's exactly how the surface subsystem works.

AntonyBoucher commented on src/platform/tiles.c:
```diff
@@ -0,0 +1,714 @@
+        SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "nearest");
+    }
+
+    // The original image will be resized to 4 possible sizes:
+    //  -  Textures[0]: tiles are   W   x   H   pixels
+    //  -  Textures[1]: tiles are (W+1) x   H   pixels
+    //  -  Textures[2]: tiles are   W   x (H+1) pixels
+    //  -  Textures[3]: tiles are (W+1) x (H+1) pixels
```
Let's just consider the width: to fill a 1920x1080 window, we would need tiles to be 19.2 pixels wide (there are 100 columns). It's not an integer.

If all tiles are the same width, then we'd use tiles that are 19 pixels wide, but that would leave 10 pixels of black padding on the left and 10 on the right. So instead, we tile the screen with 80 19px wide tile and 20 20px wide tiles, distributed evenly across the window.

You might object that we could draw the same 19-px wide glyphs into the 20-px wide tiles. That would be okay for the fitted tiles (text, monsters) but fails for the stretched ones (walls, water, etc) as you would have to choose between (1) leaving column of blank pixels between some tiles, or (2) asking the renderer to interpolate the glyph from 19 to 20 pixels, which will make it blurry.

tmewett replied:
Great explanation, thank you - could a summary of this be commented in somewhere?

AntonyBoucher commented on src/platform/tiles.c:
```diff
@@ -0,0 +1,714 @@
+    sprintf(filename, "%s/assets/tiles.png", dataDirectory);
+    SDL_Surface *image = IMG_Load(filename);
+    if (!image) imgfatal(__FILE__, __LINE__);
+    TilesPNG = SDL_ConvertSurfaceFormat(image, SDL_PIXELFORMAT_ARGB8888, 0);
```
Do you have it backwards? `ARGB888` is a *packed component order*, which means the alpha component is found in the most significant bits of a *32-bit integer*. That's consistent with the rest of my code, which treats pixels as 32-bit integers, with alpha at the top (e.g `*pixel++ = (alpha << 24) | 0xffffffU;` on line 314).

`ARGB32` is an *array order*, which treats a pixel as 4 bytes, with alpha at the lowest address. If I switched to that format, I would have to update my code to change the assumption, from "alpha in the top bits" to "alpha in the first byte".

Array formats (`RGBA32` and the likes) are only useful for serializing pixel data (e.g to disk or network), so that the components come in a predictable order regardless of machine endianness. That's why they weren't part of SDL2 until recently (2.0.5), whereas the former have been around since the beginning. For everything else, packed formats are perfectly suitable and in fact what SDL uses internally (and I believe video drivers too).

I'd be happy to switch to an array format, and modify my code accordingly, *if* someone can prove that `SDL_PIXELFORMAT_ARGB8888` performs worse than `SDL_PIXELFORMAT_BGRA8888` on big-endian machines.

tmewett replied:
Thanks for the thorough explanation - yes I did get the formats the wrong way around, my bad!

AntonyBoucher commented on src/platform/tiles.c:
```diff
@@ -0,0 +1,714 @@
+    }
+    downscaled:
+
+    // procedural floor (if the floor tile is blank)
```
Well, I do play Brogue with my own tiles, so this code is not dead to me. But I'm okay to remove it from the PR (and keep it in my branch) if you intend to discourage users to customize the tiles. Please let me know what you decide.

(Making a completely blank tile without this code kicking in is easy: just color 1 pixel a very dark grey.)

tmewett replied:
> if you intend to discourage users from customizing the tiles

It's not this at all. It's just procedural tiles like this is not really available to users as it requires game modification. The wall tops are part of the default tileset, but with all respect, it seems arbitrary to have this alternative floor tile for your use cases, I can't add everyone's.

I wonder if there's a way to meet your needs with a standard tile in the PNG? How does a noise texture look?

AntonyBoucher replied:
Okay, I removed the code and will consider it for another branch and/or PR.

AntonyBoucher commented on src/platform/tiles.c:
```diff
@@ -0,0 +1,714 @@
+    for (int x = stop0; x < stop1; x++) scaledX[x] = map0 + (map1 - map0) * (x - stop0) / (stop1 - stop0);
+    for (int x = stop1; x < stop2; x++) scaledX[x] = map1 + (map2 - map1) * (x - stop1) / (stop2 - stop1);
+    for (int x = stop2; x < stop3; x++) scaledX[x] = map2 + (map3 - map2) * (x - stop2) / (stop3 - stop2);
+    for (int x = stop3; x < stop4; x++) scaledX[x] = map3 + (map4 - map3) * (x - stop3) / (stop4 - stop3);
```
I don't understand the question. At this point in the code, we know how the left (`stop0`), right (`stop4`) and three intermediate stops (`stop1/2/3`) map to physical pixels (the `mapN` floating point coordinates), including the adjustments computed by `optimizeTiles()`. But that's only 5 pixels from the source tile, out of 128. We still have to map the coordinates of the other 123 pixels from the source (large) tile to the target (small) tile. That's what these four lines do, using linear interpolation.

tmewett replied:
I was just talking about the fact this looked like a manually-unrolled loop. But on more reading, the 0..4 vars are each significant so I retract my comment

AntonyBoucher commented on src/platform/tiles.c:
```diff
@@ -0,0 +1,714 @@
+
+
+static void loadTiles(SDL_Renderer *renderer, int outputWidth, int outputHeight) {
+    if (TilesPNG == NULL) init();
```
OK

AntonyBoucher commented:
I rebased the PR, added lots of comments into the source file, removed `xorshift64s` and `noise` (but kept the procedural floor code until you confirm), kept `ARGB8888` everywhere (since neither the code nor the format make any assumption about endianness), moved the `init` call as suggested, and remove the assumption that 1 byte = 8 bits.

tmewett commented:
> remove the assumption that 1 byte = 8 bits.

sounds crazy in retrospect :joy: oh well, `sizeof` is more readable anyway

The comments are fantastic. LGTM! Thanks very much, an exciting feature for CE!
